archive/100-theorems-list/90_stirlings_formula.lean

```diff
@@ -0,0 +1,339 @@
+open_locale classical real topological_space nnreal ennreal filter big_operators
+open  finset filter nat real
+
+namespace stirling
+
+/-- The sum of inverse squares converges. -/
+lemma summable_inverse_squares :
+summable (λ (k : ℕ), (1 : ℝ) / ((k.succ))^(2)) :=
+begin
+  have g := (summable_nat_add_iff 1).mpr (real.summable_one_div_nat_rpow.mpr one_lt_two),
+  norm_cast at *,
+  exact g,
+end
+
+/-!
+ ### Part 1
+ https://proofwiki.org/wiki/Stirling%27s_Formula#Part_1
+-/
+
+/--
+Define `stirling_seq n` as $\frac{n!}{\sqrt{2n}/(\frac{n}{e})^n$.
+Stirling's formula states that this sequence has limit $\sqrt(π)$.
+-/
+noncomputable def stirling_seq (n : ℕ) : ℝ :=
+(n.factorial : ℝ) / ((sqrt(2 * n) * ((n / (exp 1))) ^ n))
+
+/-- The function `log(1 + x) - log(1 - x)` has a power series expansion with k-th term
+`2 * x^(2 * k + 1) / (2 * k + 1)`, valid for `|x| < 1`. -/
+lemma log_sum_plus_minus (x : ℝ) (hx : |x| < 1) : has_sum (λ k : ℕ,
+  (2 : ℝ) * (1 / (2 * (k : ℝ) + 1)) * (x ^ (2 * k + 1))) (log (1 + x) - log(1 - x)) :=
+begin
+ have h₁, from has_sum_pow_div_log_of_abs_lt_1 hx,
+  have h₂, from has_sum_pow_div_log_of_abs_lt_1 (eq.trans_lt (abs_neg x) hx),
+  replace h₂ := (has_sum_mul_left_iff  (λ h : ( -1 = (0:ℝ)), one_ne_zero $ neg_eq_zero.mp h)).mp h₂,
+  rw [neg_one_mul, neg_neg, sub_neg_eq_add 1 x] at h₂,
+  have h₃, from has_sum.add h₂ h₁,
+  rw [tactic.ring.add_neg_eq_sub] at h₃,
+  let term := (λ n :ℕ, ((-1) * ((-x) ^ (n + 1) / ((n : ℝ) + 1)) + (x ^ (n + 1) / ((n : ℝ) + 1)))),
+  let two_mul := (λ (n : ℕ), (2 * n)),
+  rw ←function.injective.has_sum_iff (mul_right_injective two_pos) _ at h₃,
+  { suffices h_term_eq_goal :
+    (term ∘ two_mul) = (λ k : ℕ, 2 * (1 / (2 * (k : ℝ) + 1)) * x ^ (2 * k  + 1)),
+    by {rw h_term_eq_goal at h₃, exact h₃},
+    apply funext,
+    intro n,
+    rw [function.comp_app],
+    dsimp only [two_mul, term],
+    rw odd.neg_pow (⟨n, rfl⟩ : odd (2 * n + 1)) x,
+    rw [neg_one_mul, neg_div, neg_neg, cast_mul, cast_two],
+    ring },
+  { intros m hm,
+    rw [range_two_mul, set.mem_set_of_eq] at hm,
+    rw [even.neg_pow (even_succ.mpr hm), succ_eq_add_one, neg_one_mul, neg_add_self] },
+end
+
+/--
+For any positive real number `a`, we have the identity
+`log((a + 1) / a) = log(1 + 1 / (2*a + 1)) - log(1 - 1/(2*a + 1))`.
+-/
+lemma log_succ_div_eq_log_sub (a : ℝ) (h : 0 < a) :
+  log ((a + 1) / a) = log (1 + 1 / (2 * a + 1)) - log (1 - 1 / (2 * a + 1)) :=
+begin
+  have h₀: (2 : ℝ) * a + 1 ≠ 0, by { linarith, },
+  have h₁: a ≠ 0 := ne_of_gt h,
+  rw ← log_div,
+  suffices g, from congr_arg log g,
+  all_goals { field_simp, },
+  all_goals { linarith, },
+end
+
+/--
+For any positive real number `a`, the expression `log((a + 1) / a)` has the series expansion
+$\sum_{k=0}^{\infty}\frac{2}{2k+1}(\frac{1}{2a+1})^{2k+1}$.
+-/
+lemma power_series_log_succ_div (a : ℝ) (h : 0 < a) : has_sum (λ (k : ℕ),
+  (2 : ℝ) * (1 / (2 * (k : ℝ) + 1)) * ((1 / (2 * (a : ℝ) + 1)) ^ (2 * k + 1)))
+  (log ((a + 1) / a)) :=
```

Comment:
Move this lemma to wherever `has_sum_pow_div_log_of_abs_lt_1` lives?

Reply:
Yes, planning to do that in a seperate pull request.

Reply:
The standard thing would be to open that PR first and make this PR dependent on that PR.

Reply:
I added open the pull request for `power_series_log_succ_div` and then rebased the pull request here and added a dependency.
Not sure if there is an easy way to see the diff for the current pull request only, before a dependency pull request has been committed to master, but of course one manually select the relevant commits, e.g.:
https://github.com/leanprover-community/mathlib/pull/14874/files/a8cddf4ab4bfe95a17e54f0dc5ee6f44f6e5e293..0811000a9f6019f947b2bf5c0886d5493498e63e

Reply:
You can compare two branches: https://github.com/leanprover-community/mathlib/compare/more_log_power_series...stirling_part1